client/client.go

```diff
@@ -524,14 +524,22 @@ func (c *authenticatingClient) doAuthenticate() error {
 		err         error
 	)
 	if authDetails, err = c.authMode.Auth(c.creds); err != nil {
-		return gooseerrors.Newf(err, "authentication failed")
+		if gooseerrors.IsUnauthorised(err) {
```

Comment:
None of this is needed - the key requirement is that an error satisfying IsNotAuthorised is returned if there's an auth problem. You need to look at the Auth() implementation - the underlying issue is for example in v3Auth (and possibly elsewhere) where it does this if auth fails:
return nil, fmt.Errorf("authentication failed")

There's no indication at all that the cause was an auth error.